### ISSUE TITLE
Request for EthereumJS libraries to stop using ENV variables internally

### ISSUE THREAD

**Issue by davidmurdoch** [ethereumjs/ethereumjs-monorepo#2306]
If a user sets the `DEBUG` environment variable when using CLI tools like `ganache` and `truffle` (even if they do `DEBUG=false` or `DEBUG=0`), EthereumJS libraries take that as a signal to do their own internal debug-related things. Our CLI users most certainly don't want that. And while this can generally be worked around by CLI users, it's inconvenient for users to do so, and it leaks internal abstractions that our users don't care about.

I know Ganache could set `ejsThing.DEBUG = false;` after we instantiate, but this isn't a foolproof solution, as we would be relying on the internals of EJS package to _not_ use the flag set by `process.env.DEBUG` in its own constructor/`create`/`init`-like methods (like `await VM.create();`).

Previously we worked around this by setting `process.env.DEBUG` to `false` (to the boolean value, not as a string) when we webpacked the library; this only worked because the EJS libraries checked for truthiness (`if (this.process.env.DEBUG) {`), but now that the libraries check that the _type_ of the value of `process.env.DEBUG` is not equal to `"undefined"` our little hack doesn't work.

Anyway, we at truffle feel strongly that libraries shouldn't alter behavior based on environment variables and it'd help tremendously if EthereumJS libraries stopped listening to `process.env.DEBUG`.

**Comment by holgerd77** [comment]
Yes, that totally makes sense, I guess we haven't thought about these side effects sufficiently.

Should have no dramatic side effects if we rename to `DEBUG_EM` (for: "EthereumJS Monorepo", alternative naming suggestions welcome from everyone, should just be both expressive and short to type, not so easy 😋).

This should have no side effects on our side I would say on first guess, since we only use (and recommend this to use) for local development purposes.

I will likely plan some bug fix releases for the monorepo for next week or so with some other fixes where this could get in.

Update: just thinking, we just need to make sure that this properly works with the `debug` package in some way, will need some investigation I guess.

**Comment by ScottyPoi** [comment]
Hi @davidmurdoch I'm looking into it.
Could you provide an example for me to replicate this?

**Comment by holgerd77** [comment]
Just giving this some additional thoughts.

Maybe it is not the right approach to try to rename the DEBUG variable coming from the CLI (e.g. running something like `DEBUG=vm:tx ts-node test.ts`) - since this is very much hard-coded into the `debug` package we are using, but it might be the better approach to add an constructor option `debug` for all the packages we are using the debug package (so I guess: VM, EVM, Blockchain, maybe we can skip client for now (?)). And this `debug` option would then take precendence no matter how `DEBUG` from ENV is set.

So this code (from evm.ts e.g.):

```typescript
if (typeof process?.env.DEBUG !== 'undefined') {
      this.DEBUG = true
    }
 ```
 
 would become something like:
 ```typescript
if (options.debug !== undefinded) {
  this.debug = options.debug
} else {
  if (typeof process?.env.DEBUG !== 'undefined') {
      this.DEBUG = true
    }
}
```

Totally hacked-together pseudo-code. 😋

One main point is that the `debug` option should be optional and only taken into account if explicitly set and then ovverride the thing coming from env.

I think this would be the more robust and natural approach.

WDYGT?

**Comment by davidmurdoch** [comment]
I don't think libraries should change behavior based on environment variables, and I think the decision to avoid the overhead of `debug` calls in cases where the `debug` package won't be used was a good call (for performance reasons). If the EJS team members, or other consumers of the libraries, need the debugging behavior I think the code that initializes these classes should be responsible for reading from the environment and setting the `debug` option, just as you've prototyped (I just prefer to drop the `process.env.DEBUG` check entirely).

If it is decided that EJS libraries must continue to branch off environment variables, you can do this with finer granularity via the `debug` module itself, e.g.: 

```javascript
if (debug.enabled("evm") ) {
  const { caller, gasLimit, to, value, delegatecall } = message
  debug(
    `New message caller=${caller} gasLimit=${gasLimit} to=${
      to?.toString() ?? 'none'
     } value=${value} delegatecall=${delegatecall ? 'yes' : 'no'}`
  )
}
```

Tangentially related: you could also lazily initialize the `debug` module; only creating the logger if `options.debug` was enabled.

**Comment by ScottyPoi** [comment]
`debug` module has a `disable()` method, so i think it could be as simple as this:

```
if (!options.debug) {
   debug.disable()
  }
```

it should effectively work the same as a conditional at each `debug()` call.  The `debug()` statements will be skipped unless `debug: true` is set in the constructor.

**Comment by ScottyPoi** [comment]
Or more explicitly:
```
    if (opts.debug === true) {
      createDebugLogger.enable('evm, evm:gas')
      this.DEBUG = true
    } else {
      createDebugLogger.disable()
    }
```
This would override the process.env.DEBUG setting though.  If the class were instantiated with `debug=true`, you would not be able to silence it with `DEBUG=0` or `DEBUG=false`.

**Comment by davidmurdoch** [comment]
If you don't have the `if (this.debug) { debug( ... ); }` check, and instead just directly do `debug(...)`, relying on the previous call to `enable`/`disable` the `debug` statements won't be "skipped" -- they just won't cause the `debug` module to log. The problem with this is that we'll still incur the overhead of creating the strings passed to `debug`, which was the reason the `this.debug` checks were added in the first place, as the performance of applications was non-trivially impacted, IIRC.

**Comment by ScottyPoi** [comment]
Okay I can leave in the individual `if` statements in there.

>  ...you could also lazily initialize the debug module; only creating the logger if options.debug was enabled.

Will do this as well.

One last question is whether to include something like:
```
if (process.env.DEBUG === "0" || process.env.DEBUG === "false") {
     debug.disable()
     }
```

Which would allow for the debugger to be silenced even if `debug: true` is set.

**Comment by jochem-brouwer** [comment]
Closed by #2433 

@davidmurdoch can you verify? Please re-open if this did not fix the problem!
